pageserver/src/tenant/storage_layer/image_layer.rs

```diff
@@ -550,15 +551,22 @@ impl ImageLayerInner {
 
             let buf = BytesMut::with_capacity(buf_size);
             let blobs_buf = vectored_blob_reader.read_blobs(&read, buf, ctx).await?;
-
             let frozen_buf = blobs_buf.buf.freeze();
 
             for meta in blobs_buf.blobs.iter() {
                 let img_buf = frozen_buf.slice(meta.start..meta.end);
+                let decompressed =
+                    vectored_blob_io::decompress(&img_buf, meta.compression_bits).await?;
+
+                let img_buf = if let Some(decompressed) = decompressed {
+                    decompressed
+                } else {
+                    img_buf
+                };
 
                 key_count += 1;
                 writer
-                    .put_image(meta.meta.key, img_buf, ctx)
+                    .put_image(meta.meta.key, Bytes::from(Vec::from(img_buf)), ctx)
```

Comment:
Doesn't this create an unnecessary copy of the img_buf?

`Vec::from` from slice will always allocate, bytes conversion from vec could allocate.

It seems earlier it was fine to pass it as a slice to `put_image` but does it now need to be wrapped as Bytes?

If slice is fine, then you can just use `decompressed.as_deref().unwrap_or(slice)` again.

Reply:
```suggestion
                    .put_image(meta.meta.key, img_buf, ctx)
```

Reply:
missed this. It now looks like

https://github.com/neondatabase/neon/blob/1ca519fbb33e4e231dd529a9ab7d06495f80f315/pageserver/src/tenant/storage_layer/image_layer.rs#L558-L559